### DISCUSSION TITLE
Intermittent 401s and high-number of active connections

### DISCUSSION THREAD

**Discussion by MarkLambeNitro** [DependencyTrack/dependency-track#2409] [Q&A] [answered]
Hi All, I'm a bit confused as to what might be causing this problem and would love any insight you might have.

I've deployed dependency-track to our K8s cluster, API and Front-end on the same machine as we're just testing it out. The Database is on a separate (RDS, Postgres) machine. Both machines have >4B RAM / 2vCPUs so I don't think it's a resource issue. They're both using a small % of available resources. 

I can log in as admin. I can create users. However, about 90% of the time the API calls will return a 401. So a specific API will work and then fail with 401 a second or two later. Also, if that failure doesn't log me out, another API call might work fine. Obviously it does result in me getting logged out quite quickly.

There's no one else using this service, just me. I created another user and get the same behaviour. 

I've noticed that the database often has > 30 connections open from the API, however, even restarting the machines doesn't fix this, they just open again. Killing them manually similarly has no affect. 

Happy to provide any other information that might be useful. I'm essentially running with default settings, haven't changed anything other than telling the API where to find the DB. 

Thanks!

Edit: I thought perhaps it's the CA cert as we're behind cloudflare but that wouldn't be intermittent right and should show a different error.

**Comment by nscuro** [comment, marked as answer]
Are you able to confirm that the requests resulting in a 401 response actually do reach the API server container?

I have not yet seen any cases where resource exhaustion caused 401s. 
Out of interest: Did you deploy multiple replicas of the API server? Could this be due to load balancing?

Regarding the DB connections, it is indeed a bit counter-intuitive. The API server actually maintains *two* connection pools. This has to do with how the ORM we're using works. This is documented here: https://docs.dependencytrack.org/getting-started/database-support/#connection-pooling

For your small-ish deployment, you can simply reduce the connection pool sizes to something like this:
* `ALPINE_DATABASE_POOL_MAX_SIZE=10` (or `ALPINE_DATABASE_POOL_MAX_SIZE=5`)
* `ALPINE_DATABASE_POOL_MIN_IDLE=5` (or `ALPINE_DATABASE_POOL_MIN_IDLE=0`)

If you can, make use of the Prometheus metrics we're exposing to observe whether the configured pool sizes work for your deployment.

**Reply by MarkLambeNitro** [reply]
> Out of interest: Did you deploy multiple replicas of the API server? Could this be due to load balancing?

Oh man I can't believe I didn't think of that, we deploy two by default. Brought it down to one machine and it works perfectly. You've just saved me a load of time, thanks!
